src/vmware/vmxnet3_queue.c

```diff
@@ -0,0 +1,404 @@
+#include "vmxnet3_queue.h"
+#include "vmxnet3_net.h"
+#include "netif/ethernet.h"
+
+#ifdef VMXNET3_NET_DEBUG
+# define vmxnet3_net_debug rprintf
+#else
+# define vmxnet3_net_debug(...) do { } while(0)
+#endif // defined(VMXNET3_NET_DEBUG)
+
+static void
+vmxnet3_init_txq(vmxnet3_pci dev, int q)
+{
+    struct vmxnet3_txqueue *txq;
+    struct vmxnet3_comp_ring *txc;
+    struct vmxnet3_txring *txr;
+
+    txq = dev->vmx_txq[q];
+    txc = &txq->vxtxq_comp_ring;
+    txr = &txq->vxtxq_cmd_ring;
+
+    buffer b = little_stack_buffer(16);
+    bprintf(b, "%s-tx%d", DEVICE_NAME, q);
+    memcpy(txq->vxtxq_name, b->contents, sizeof(txq->vxtxq_name));
+
+    txq->vxtxq_sc = dev;
+    txq->vxtxq_id = q;
+    txc->vxcr_ndesc = VMXNET3_MAX_TX_NCOMPDESC;
+    txr->vxtxr_ndesc = VMXNET3_MAX_TX_NDESC;
+    txr->vxtxr_avail = VMXNET3_MAX_TX_NDESC;
+    txr->vxtxr_head = 0;
+}
+
+static void
+vmxnet3_init_rxq(vmxnet3_pci dev, int q)
+{
+    struct vmxnet3_rxqueue *rxq;
+    struct vmxnet3_comp_ring *rxc;
+    struct vmxnet3_rxring *rxr;
+
+    rxq = dev->vmx_rxq[q];
+    rxc = &rxq->vxrxq_comp_ring;
+
+    buffer b = little_stack_buffer(16);
+    bprintf(b, "%s-rx%d", DEVICE_NAME, q);
+    memcpy(rxq->vxrxq_name, b->contents, sizeof(rxq->vxrxq_name));
+
+    rxq->vxrxq_sc = dev;
+    rxq->vxrxq_id = q;
+
+    rxc->vxcr_ndesc = VMXNET3_MAX_RX_NCOMPDESC;
+    for (int i = 0; i < VMXNET3_RXRINGS_PERQ; i++) {
+        rxr = &rxq->vxrxq_cmd_ring[i];
+        rxr->vxrxr_ndesc = VMXNET3_MAX_RX_NDESC;
+    }
+}
+
+void
+init_vmxnet3_tx_queue(vmxnet3_pci vp, struct vmxnet3_txqueue *txq)
+{
+    struct vmxnet3_txring *txr = &txq->vxtxq_cmd_ring;
+
+    txq->vxtxq_last_flush = -1;
+
+    txr = &txq->vxtxq_cmd_ring;
+    txr->vxtxr_next = 0;
+    txr->vxtxr_gen = VMXNET3_INIT_GEN;
+
+    struct vmxnet3_comp_ring *txc = &txq->vxtxq_comp_ring;
+    txc->vxcr_next = 0;
+    txc->vxcr_gen = VMXNET3_INIT_GEN;
+}
+
+void init_vmxnet3_rx_queue(vmxnet3_pci vp, struct vmxnet3_rxqueue *rxq)
+{
+    for (int i=0; i < VMXNET3_RXRINGS_PERQ; i++) {
+        struct vmxnet3_rxring *rxr = &rxq->vxrxq_cmd_ring[i];
+        rxr->vxrxr_gen = VMXNET3_INIT_GEN;
+        rxr->vxrxr_desc_skips = 0;
+        rxr->vxrxr_refill_start = 0;
+    }
+
+    struct vmxnet3_comp_ring *rxc = &rxq->vxrxq_comp_ring;
+    rxc->vxcr_next = 0;
+    rxc->vxcr_gen = VMXNET3_INIT_GEN;
+    rxc->vxcr_zero_length = 0;
+    rxc->vxcr_pkt_errors = 0;
+}
+
+void vmxnet3_tx_queues_alloc(vmxnet3_pci dev)
+{
+    for (int i = 0; i < VMXNET3_DEF_TX_QUEUES; ++i) {
+        dev->vmx_txq[i] = allocate_zero(dev->contiguous, sizeof(struct vmxnet3_txqueue));
+        assert(dev->vmx_txq[i] != INVALID_ADDRESS);
+        vmxnet3_init_txq(dev, i);
+        init_vmxnet3_tx_queue(dev, dev->vmx_txq[i]);
+    }
+
+    // allocate tx descriptors memory
+    u64 tx_desc_size = sizeof(struct vmxnet3_txdesc) * VMXNET3_MAX_TX_NDESC * VMXNET3_DEF_TX_QUEUES;
+    dev->tx_desc_mem = allocate_zero(dev->contiguous, tx_desc_size);
```

Comment:
again, include space for padding? plus four instances below...

Reply:
There is an assert() to check that we have correct alignment - I thought it should be enough